test/core/security/tls_security_connector_test.cc

```diff
@@ -376,6 +376,52 @@ TEST_F(TlsSecurityConnectorTest, TlsCheckHostNameFail) {
   GRPC_ERROR_UNREF(error);
 }
 
+TEST_F(TlsSecurityConnectorTest, CompareChannelSecurityConnectorSucceeds) {
+  grpc_core::RefCountedPtr<grpc_tls_certificate_distributor> distributor =
+      grpc_core::MakeRefCounted<grpc_tls_certificate_distributor>();
+  distributor->SetKeyMaterials(kRootCertName, root_cert_0_, absl::nullopt);
+  grpc_core::RefCountedPtr<::grpc_tls_certificate_provider> provider =
+      grpc_core::MakeRefCounted<TlsTestCertificateProvider>(distributor);
+  auto options = grpc_core::MakeRefCounted<grpc_tls_credentials_options>();
+  options->set_certificate_provider(provider);
+  options->set_watch_root_cert(true);
+  options->set_root_cert_name(kRootCertName);
+  grpc_core::RefCountedPtr<TlsCredentials> credential =
+      grpc_core::MakeRefCounted<TlsCredentials>(options);
+  grpc_core::RefCountedPtr<grpc_channel_security_connector> connector =
+      credential->create_security_connector(nullptr, kTargetName, nullptr,
+                                            nullptr);
+  EXPECT_EQ(connector->cmp(connector.get()), 0);
+}
+
+TEST_F(TlsSecurityConnectorTest, CompareChannelSecurityConnectorFails) {
```

Comment:
I think we need to test that all fields are covered here.  For each field, we should create two security connectors that differ by only that field, and then verify that they are not equal.

Reply:
Sure, I can add that. Just to make sure, since we already have tests covering two security connectors with same option settings that would fail, do we still want to make one field differ and test them?
I am happy to add those tests, but just from my point of view, if two connectors are already different, there is no chance for  the compare function to return true, since their memory locations are already different. Maybe we want to test one connector with different settings, to indicate the fact that changing the state of options would actually result in a different security connector?

Reply:
Thanks for pointing this out!

This comment made me realize that most of this PR is actually unnecessary.  I had forgotten about the fact that the SC comparison returns false if the SCs are from different credentials objects, even if they have the same options.  Given that, there's no need to explicitly check the options in the SC comparison function, because two SCs from the same credential object will always have the same options.

I think we can remove all of the code changes and just leave the tests here to verify the right behavior.  Let's just change this test to use the same options object for the two credentials objects, just to make it really clear that it's the credentials instance that matters, not the options instance.

Sorry for steering you wrong here!

Reply:
No problem! I think this also means https://github.com/grpc/grpc/pull/25631 is not blocked by this anymore. I updated the tests to reflect the latest changes. Thanks!